### PR TITLE
Handle vector width abi

### PR DESCRIPTION
**Pull request by wsmoses**
Requires https://github.com/EnzymeAD/Enzyme/pull/615

### CODE REVIEW

File: src/Enzyme.jl [hunk 1]
```diff
@@ -69,6 +69,28 @@ struct DuplicatedNoNeed{T} <: Annotation{T}
...
+Mark a function argument `x` of [`autodiff`](@ref) as duplicated, Enzyme will
+auto-differentiate in respect to such arguments, with `dx` acting as an
+accumulator for gradients (so ``\\partial f / \\partial x`` will be *added to*)
+`∂f_∂x`.
+"""
+struct BatchDuplicated{T,N} <: Annotation{T}
+    val::T
+    dval::NTuple{N,T}
+end
+batch_size(::BatchDuplicated{T,N}) where {T,N} = N
```
Comment (vchuravy) [inline src/Enzyme.jl:87]:
```suggestion
batch_size(::BatchDuplicated{T,N}) where {T,N} = N

```

File: src/Enzyme.jl [hunk 2]
```diff
@@ -116,10 +138,30 @@ import .Compiler: CompilationException
     end
 end
 
-prepare_cc() = ()
-prepare_cc(arg::Duplicated, args...) = (arg.val, arg.dval, prepare_cc(args...)...)
-prepare_cc(arg::DuplicatedNoNeed, args...) = (arg.val, arg.dval, prepare_cc(args...)...)
-prepare_cc(arg::Annotation, args...) = (arg.val, prepare_cc(args...)...)
+@inline function same_or_one(args::Vararg{Any, N}) where N
```
Comment (vchuravy) [inline src/Enzyme.jl:141]: Is this fully eliminated? Alternative recursive implementation probable is nicer on the compiler

File: src/Enzyme.jl [hunk 3]
```diff
@@ -470,4 +518,59 @@ macro parallel(args...)
    end)
 end
 
+@inline function onehot(x, start=1, endl=length(x))
+    ntuple(Val(endl-start+1)) do i
```
Comment (vchuravy) [inline src/Enzyme.jl:522]: Where did you get this implementation from xD? Drop the `Val`?

File: src/compiler.jl [hunk 4]
```diff
@@ -1242,37 +1300,85 @@ end
...
+        end
+        to_preserve = LLVM.Value[vals[1], vals[2]]
+        T_args = LLVM.LLVMType[T_prjlvalue, T_prjlvalue]
+    else
+        to_preserve = LLVM.Value[]
+        T_args = LLVM.LLVMType[]
+    end
+@static if VERSION < v"1.8-"
+else
+        push!(vals, LLVM.Value(API.EnzymeGradientUtilsNewFromOriginal(gutils, ops[end])))
```
Comment (vchuravy) [inline src/compiler.jl:1366]:
```suggestion
    push!(vals, LLVM.Value(API.EnzymeGradientUtilsNewFromOriginal(gutils, ops[end])))
```

File: src/compiler.jl [hunk 5]
```diff
@@ -3694,57 +3840,122 @@ function _thunk(job)
...
+            cache[key] = obj
+        end
+        obj
+    finally
+        unlock(cache_lock)
+    end
+end
+
+
+@generated function genthunk(f::F, df::DF, ::Type{A}, tt::Type{TT},::Val{Mode}, ::Val{width}, ::Val{specid}) where {F, DF, A<:Annotation, TT, Mode, width, specid}
```
Comment (vchuravy) [inline src/compiler.jl:3870]:
```suggestion
@generated function __thunk(f::F, df::DF, ::Type{A}, tt::Type{TT},::Val{Mode}, ::Val{width}, ::Val{specid}) where {F, DF, A<:Annotation, TT, Mode, width, specid}
```

File: src/compiler.jl [hunk 6]
```diff
@@ -3694,57 +3840,122 @@ function _thunk(job)
...
+@inline function thunk(f::F,df::DF, ::Type{A}, tt::Type{TT},::Val{Mode}, ::Val{width}) where {F, DF, A<:Annotation, TT, Mode, width}
+    primal, adjoint = fspec(f, tt)
+
+    target = Compiler.EnzymeTarget()
+    params = Compiler.EnzymeCompilerParams(adjoint, Mode, width, A, true, DF != Nothing, #=abiwrap=#true)
+    job    = Compiler.CompilerJob(target, primal, params)
+
+    specid = GPUCompiler.specialization_id(job)
+
+    genthunk(f, df, A, TT, Val(Mode), Val(width), Val(specid))
```
Comment (vchuravy) [inline src/compiler.jl:3949]:
```suggestion
    __thunk(f, df, A, TT, Val(Mode), Val(width), Val(specid))
```

File: src/compiler/reflection.jl [hunk 7]
```diff
@@ -2,12 +2,13 @@ function get_job(@nospecialize(func), @nospecialize(A), @nospecialize(types);
                  run_enzyme::Bool=true, mode::API.CDerivativeMode=API.DEM_ReverseModeCombined, dupClosure::Bool=false, argwrap::Bool=true, kwargs...)
 
     primal, adjoint = fspec(func, types)
+    width = 1
```
Comment (vchuravy) [inline src/compiler/reflection.jl:5]: Turn this into a kwarg?

File: test/runtests.jl [hunk 8]
```diff
@@ -835,3 +835,25 @@ end
...
 end
+
+@testset "Batch" begin
+    square(x)=x*x
+    bres = fwddiff(square, BatchDuplicatedNoNeed, BatchDuplicated(3.0, (1.0, 2.0, 3.0)))
+    @test length(bres) == 1
+    @test length(bres[1]) == 3
+    @test bres[1] ≈ (6.0, 12.0, 18.0),
+
+    bres = fwddiff(square, BatchDuplicatedNoNeed, BatchDuplicated(3.0 + 7.0im, (1.0+0im, 2.0+0im, 3.0+0im)))
```
Comment (vchuravy) [inline test/runtests.jl:846]: I am still against `Duplicated` for immutables. @ZuseZ4 proposed `Dual` the other day as well

**Review by vchuravy — APPROVED** [review]
Modulo removal of debug output and test passing

**Review by tgymnich — APPROVED** [review]

File: src/Enzyme.jl [hunk 9]
```diff
@@ -582,7 +589,52 @@ end
...
+        return dx
+    end
+    tupleconcat(tmp...)
+end
+
+@inline function revjacobian(f, x, ::Val{1}; n_outs::Val{n_out_val}) where {n_out_val}
+    tt′    = Tuple{Duplicated{Core.Typeof(x)}}
+    tt    = Tuple{Core.Typeof(x)}
+    rt = Core.Compiler.return_type(f, tt)
+    forward, adjoint = Enzyme.Compiler.thunk(f, #=df=#nothing, DuplicatedNoNeed{rt}, tt′, #=Split=# Val(API.DEM_ReverseModeGradient), #=width=#Val(1), #=ModifiedBetween=#Val(false))
```
Comment (vchuravy) [inline src/Enzyme.jl:628]:
```suggestion
    primal, adjoint = Enzyme.Compiler.thunk(f, #=df=#nothing, DuplicatedNoNeed{rt}, tt′, #=Split=# Val(API.DEM_ReverseModeGradient), #=width=#Val(1), #=ModifiedBetween=#Val(false))
```

File: src/Enzyme.jl [hunk 10]
```diff
@@ -582,7 +589,52 @@ end
...
 
-# x  = [Float64(i) for i in 1:10]; x2 = ntuple(10) do i x[i] end;
-#
+@inline function revjacobian(f, x, ::Val{chunk}; n_outs::Val{n_out_val}) where {chunk, n_out_val}
+    num = ((n_out_val + chunk - 1) ÷ chunk)
+
+    tt′    = Tuple{BatchDuplicated{Core.Typeof(x), chunk}}
+    tt    = Tuple{Core.Typeof(x)}
+    rt = Core.Compiler.return_type(f, tt)
+    forward, adjoint = Enzyme.Compiler.thunk(f, #=df=#nothing, BatchDuplicatedNoNeed{rt}, tt′, #=Split=# Val(API.DEM_ReverseModeGradient), #=width=#Val(chunk), #=ModifiedBetween=#Val(false))
```
Comment (vchuravy) [inline src/Enzyme.jl:598]:
```suggestion
    primal, adjoint = Enzyme.Compiler.thunk(f, #=df=#nothing, BatchDuplicatedNoNeed{rt}, tt′, #=Split=# Val(API.DEM_ReverseModeGradient), #=width=#Val(chunk), #=ModifiedBetween=#Val(false))
```

File: src/compiler.jl [hunk 11]
```diff
@@ -2779,14 +2782,14 @@ function enzyme!(job, mod, primalf, adjoint, mode, width, parallel, actualRetTyp
     typeInfo = FnTypeInfo(retTT, args_typeInfo, args_known_values)
 
     if mode == API.DEM_ReverseModePrimal || mode == API.DEM_ReverseModeGradient
-        returnUsed = !(GPUCompiler.isghosttype(actualRetType) || Core.Compiler.isconstType(actualRetType))
+        returnUsed = !(GPUCompiler.isghosttype(actualRetType) || Core.Compiler.isconstType(actualRetType)) 
         shadowReturnUsed = returnUsed && (retType == API.DFT_DUP_ARG || retType == API.DFT_DUP_NONEED)
+        returnUsed &= retType != API.DFT_DUP_NONEED
         # TODO handle non-1 width
```
Comment (vchuravy) [inline src/compiler.jl:2788]:
```suggestion
```

File: src/compiler.jl [hunk 12]
```diff
@@ -2918,7 +2921,7 @@ function create_abi_wrapper(enzymefn::LLVM.Function, F, argtypes, rettype, actua
         llvmT = isboxed ? T_prjlvalue : convert(LLVMType, actualRetType; ctx)
         
         # primal return
-        if existed[2] != 0
+        if existed[2] != 0 &&
```
Comment (vchuravy) [inline src/compiler.jl:2924]:
```suggestion
        if existed[2] != 0
```

File: src/compiler.jl [hunk 13]
```diff
@@ -4012,24 +4016,24 @@ end
     end
 end
 
-@inline function thunk(f::F,df::DF, ::Type{A}, tt::Type{TT},::Val{Mode}, ::Val{width}) where {F, DF, A<:Annotation, TT, Mode, width}
+@inline function thunk(f::F,df::DF, ::Type{A}, tt::Type{TT},::Val{Mode}, ::Val{width}, ::Val{ModifiedBetween}=Val(Mode!= API.DEM_ReverseModeCombined)) where {F, DF, A<:Annotation, TT, Mode, width, ModifiedBetween}
```
Comment (vchuravy) [inline src/compiler.jl:4019]:
```suggestion
@inline function thunk(f::F,df::DF, ::Type{A}, tt::Type{TT},::Val{Mode}, ::Val{width}, ::Val{ModifiedBetween}=Val(Mode != API.DEM_ReverseModeCombined)) where {F, DF, A<:Annotation, TT, Mode, width, ModifiedBetween}
```

File: src/compiler.jl [hunk 14]
```diff
@@ -4012,24 +4016,24 @@ end
...
 
-    genthunk(f, df, A, TT, Val(Mode), Val(width), Val(specid))
+    genthunk(f, df, A, TT, Val(Mode), Val(ModifiedBetween), Val(width), Val(specid))
 end
 
 import GPUCompiler: deferred_codegen_jobs
 
 @generated function deferred_codegen(f::F, ::Val{tt}, ::Val{rt}, ::Val{DupClosure},::Val{Mode},
-                                     ::Val{width}) where {F,tt, rt, DupClosure, Mode, width}
+                                     ::Val{width}, ::Val{ModifiedBetween}=Val(Mode!= API.DEM_ReverseModeCombined)) where {F,tt, rt, DupClosure, Mode, width, ModifiedBetween}
```
Comment (vchuravy) [inline src/compiler.jl:4033]:
```suggestion
                                     ::Val{width}, ::Val{ModifiedBetween}=Val(Mode != API.DEM_ReverseModeCombined)) where {F,tt, rt, DupClosure, Mode, width, ModifiedBetween}
```

File: src/Enzyme.jl [hunk 15]
```diff
@@ -549,8 +549,13 @@ end
 end
 
 @inline function revgradient(f, x)
-    dx = zeros(x)
-    fwddiff(f, Duplicated(x, dx))
+    dx = zero(x)
+    autodiff(f, Duplicated(x, dx))
+    dx
+end
```
Comment (vchuravy) [inline src/Enzyme.jl:555]:
```suggestion
end

```

File: src/Enzyme.jl [hunk 16]
```diff
@@ -191,12 +242,13 @@ while ``\\partial f/\\partial b`` will be *added to* `∂f_∂b` (but not return
...
         end
-    elseif A <: Duplicated
+    elseif A <: Duplicated || A<: DuplicatedNoNeed || A <: BatchDuplicated || A<: BatchDuplicatedNoNeed
         throw(ErrorException("Duplicated Returns not yet handled"))
     end
-    thunk = Enzyme.Compiler.thunk(f, #=df=#nothing, A, tt′, #=Split=# Val(API.DEM_ReverseModeCombined))
-    rt = eltype(Compiler.return_type(thunk))
+    thunk = Enzyme.Compiler.thunk(f, #=df=#nothing, A, tt′, #=Split=# Val(API.DEM_ReverseModeCombined), width)
     if A <: Active
+        tt    = Tuple{map(T->eltype(Core.Typeof(T)), args′)...}
```
Comment (vchuravy) [inline src/Enzyme.jl:250]:
```suggestion
        tt = Tuple{map(T->eltype(Core.Typeof(T)), args′)...}
```

File: src/Enzyme.jl [hunk 17]
```diff
@@ -470,4 +527,124 @@ macro parallel(args...)
...
+            @inbounds shadow[(i-1)*chunk+j] += one(eltype(typeof(shadow)))
+        end
+        (i == num ? adjoint2 : adjoint)(BatchDuplicated(x, dx), tape)
+        return dx
+    end
+    tupleconcat(tmp...)
+end
+
+@inline function revjacobian(f, x, ::Val{1}; n_outs::Val{n_out_val}) where {n_out_val}
+    tt′    = Tuple{Duplicated{Core.Typeof(x)}}
```
Comment (vchuravy) [inline src/Enzyme.jl:635]: formatting

File: src/api.jl [hunk 18]
```diff
@@ -142,14 +143,14 @@ end
 #  \p PostOpt is whether to perform basic optimization of the function after synthesis
 function EnzymeCreateAugmentedPrimal(logic, todiff, retType, constant_args, TA,  returnUsed,
                                      shadowReturnUsed,
-                                     typeInfo, uncacheable_args, forceAnonymousTape, atomicAdd)
+                                     typeInfo, uncacheable_args, forceAnonymousTape, width, atomicAdd)
     ccall((:EnzymeCreateAugmentedPrimal, libEnzyme), EnzymeAugmentedReturnPtr, 
         (EnzymeLogicRef, LLVMValueRef, CDIFFE_TYPE, Ptr{CDIFFE_TYPE}, Csize_t, 
          EnzymeTypeAnalysisRef, UInt8, UInt8, 
-         CFnTypeInfo, Ptr{UInt8}, Csize_t, UInt8, UInt8),
+         CFnTypeInfo, Ptr{UInt8}, Csize_t, UInt8, Cuint, UInt8),
```
Comment (vchuravy) [inline src/api.jl:150]: I really wish enzyme proper would have a ABI query....

File: src/api.jl [hunk 19]
```diff
@@ -186,6 +187,10 @@ const CustomReversePass = Ptr{Cvoid}
 EnzymeRegisterCallHandler(name, fwdhandle, revhandle) = ccall((:EnzymeRegisterCallHandler, libEnzyme), Cvoid, (Cstring, CustomAugmentedForwardPass, CustomReversePass), name, fwdhandle, revhandle)
 EnzymeRegisterFwdCallHandler(name, fwdhandle) = ccall((:EnzymeRegisterFwdCallHandler, libEnzyme), Cvoid, (Cstring, CustomForwardPass), name, fwdhandle)
 
+EnzymeGetShadowType(width, T) = ccall((:EnzymeGetShadowType, libEnzyme), LLVMTypeRef, (UInt64,LLVMTypeRef), width, T)
+
+
```
Comment (vchuravy) [inline src/api.jl:192]:
```suggestion
```

File: src/Enzyme.jl [hunk 20]
```diff
@@ -470,4 +527,124 @@ macro parallel(args...)
...
+    ntuple(length(shadow)) do i
+        fwddiff(f, DuplicatedNoNeed, Duplicated(x, shadow[i]))[1]
+    end
+end
+
+@inline function fwdjacobian(args...; kwargs...)
+    fwdgradient(args...; kwargs...)
+end
+
+@inline function revjacobian(f, x, ::Val{chunk}; n_outs::Val{n_out_val}) where {chunk, n_out_val}
```
Comment (vchuravy) [inline src/Enzyme.jl:598]: I would appreciate some comments.

File: src/compiler.jl [hunk 21]
```diff
@@ -951,85 +964,100 @@ end
...
     orig = LLVM.Instruction(OrigCI)
-    normal = (unsafe_load(normalR) != C_NULL) ? LLVM.Instruction(unsafe_load(normalR)) : nothing
-    shadow = (unsafe_load(shadowR) != C_NULL) ? LLVM.Instruction(unsafe_load(shadowR)) : nothing
-    mod = LLVM.parent(LLVM.parent(LLVM.parent(orig)))
-    ctx = LLVM.context(orig)
 
-    conv = LLVM.API.LLVMGetInstructionCallConv(orig)
-    # https://github.com/JuliaLang/julia/blob/5162023b9b67265ddb0bbbc0f4bd6b225c429aa0/src/codegen_shared.h#L20
-    @assert conv == 38
+    if API.EnzymeGradientUtilsIsConstantValue(gutils, orig) == 0 || API.EnzymeGradientUtilsIsConstantInstruction(gutils, orig) == 0
```
Comment (vchuravy) [inline src/compiler.jl:968]: we should use a more Julian API

File: src/compiler.jl [hunk 22]
```diff
@@ -1182,24 +1229,43 @@ end
...
     # the original compilation (rather than runtime JIT compiling)
-    forward, adjoint = thunk(func, dfunc, Const, tt, Val(API.DEM_ReverseModePrimal))
+    forward = thunk(func, dfunc, Const, tt, Val(API.DEM_ForwardMode), width)
 
-	tapes = Vector{Ptr{Cvoid}}(undef, Base.Threads.nthreads())
+    function fwd()
+        forward()
+    end
+
+    Base.Threads.threading_run(fwd)
```
Comment (vchuravy) [inline src/compiler.jl:1247]:
```suggestion
    Base.Threads.threading_run(forward)
```
?

File: src/compiler.jl [hunk 23]
```diff
@@ -1210,28 +1276,45 @@ end
...
     # the original compilation (rather than runtime JIT compiling)
-    forward, adjoint = thunk(func, dfunc, Const, tt, Val(API.DEM_ReverseModePrimal))
+    forward = thunk(func, dfunc, Const, tt, Val(API.DEM_ForwardMode), width)
 
-	tapes = Vector{Ptr{Cvoid}}(undef, Base.Threads.nthreads())
+    function fwd()
+        forward()
+    end
+
+    Base.Threads.threading_run(fwd, dynamic)
```
Comment (vchuravy) [inline src/compiler.jl:1296]:
```suggestion
    Base.Threads.threading_run(forward, dynamic)
```

File: src/compiler.jl [hunk 24]
```diff
@@ -1210,28 +1276,45 @@ end
...
+    tapes = Vector{Ptr{Cvoid}}(undef, Base.Threads.nthreads())
     # tapes = unsafe_convert(Ptr{Ptr{Cvoid}}, Libc.malloc(sizeof(Ptr{Cvoid})*Base.Threads.nthreads))
-	function fwd()
-		tapes[Base.Threads.threadid()] = forward()[1]
-	end
-	adj = () -> adjoint(tapes[Base.Threads.threadid()])
+    function fwd(tid)
+        tapes[tid] = forward()[1]
+    end
+    adj = (tid) -> adjoint(tapes[tid])
```
Comment (vchuravy) [inline src/compiler.jl:1314]: Are these threading fixes for 1.8? Split them out into a separate PR we will need to backport them

File: src/compiler.jl [hunk 25]
```diff
@@ -1640,14 +1791,32 @@ function arraycopy_common(fwd, B, orig, origArg, gutils)
...
+    width = API.EnzymeGradientUtilsGetWidth(gutils)
+    if width == 1
+    
     shadowsrc = load!(B, bitcast!(B, shadowsrc, LLVM.PointerType(LLVM.PointerType(LLVM.IntType(8; ctx), 13), LLVM.addrspace(LLVM.llvmtype(shadowsrc)))))
+    shadowdst = load!(B, bitcast!(B, shadowdst, LLVM.PointerType(LLVM.PointerType(LLVM.IntType(8; ctx), 13), LLVM.addrspace(LLVM.llvmtype(shadowdst)))))
 
     API.EnzymeGradientUtilsSubTransferHelper(gutils, fwd ? API.DEM_ReverseModePrimal : API.DEM_ReverseModeGradient, secretty, intrinsic, #=dstAlign=#1, #=srcAlign=#1, #=offset=#0, false, shadowdst, false, shadowsrc, length, isVolatile, orig, allowForward, #=shadowsLookedUp=#!fwd)
+    
+    else
+    for i in 1:width
```
Comment (vchuravy) [inline src/compiler.jl:1808]: formatting

File: src/compiler.jl [hunk 26]
```diff
@@ -1640,14 +1791,32 @@ function arraycopy_common(fwd, B, orig, origArg, gutils)
...
         shadowdst = LLVM.Value(API.EnzymeGradientUtilsLookup(gutils, shadowdst, B))
     end
-    shadowdst = load!(B, bitcast!(B, shadowdst, LLVM.PointerType(LLVM.PointerType(LLVM.IntType(8; ctx), 13), LLVM.addrspace(LLVM.llvmtype(shadowdst)))))
     shadowsrc = LLVM.Value(API.EnzymeGradientUtilsInvertPointer(gutils, origArg, B))
     if !fwd
         shadowsrc = LLVM.Value(API.EnzymeGradientUtilsLookup(gutils, shadowsrc, B))
     end
+    
+    width = API.EnzymeGradientUtilsGetWidth(gutils)
+    if width == 1
```
Comment (vchuravy) [inline src/compiler.jl:1800]: formatting

File: src/compiler.jl [hunk 27]
```diff
@@ -1674,19 +1855,96 @@ function arraycopy_rev(B::LLVM.API.LLVMBuilderRef, OrigCI::LLVM.API.LLVMValueRef
...
+    return nothing
+end
 
+function f_tuple_augfwd(B::LLVM.API.LLVMBuilderRef, OrigCI::LLVM.API.LLVMValueRef, gutils::API.EnzymeGradientUtilsRef, normalR::Ptr{LLVM.API.LLVMValueRef}, shadowR::Ptr{LLVM.API.LLVMValueRef}, tapeR::Ptr{LLVM.API.LLVMValueRef})::Cvoid
+    f_tuple_fwd(B, OrigCI, gutils, normalR, shadowR)
     return nothing
 end
 
 function f_tuple_rev(B::LLVM.API.LLVMBuilderRef, OrigCI::LLVM.API.LLVMValueRef, gutils::API.EnzymeGradientUtilsRef, tape::LLVM.API.LLVMValueRef)::Cvoid
-    emit_error(LLVM.Builder(B), "Enzyme: Not yet implemented reverse for jl_f_tuple")
```
Comment (vchuravy) [inline src/compiler.jl:1689]: Add a comment why this is nothing.

File: src/compiler.jl [hunk 28]
```diff
@@ -2375,11 +2919,14 @@ function int_return_rule(direction::Cint, ret::API.CTypeTreeRef, args::Ptr{API.C
 end
 
 function i64_box_rule(direction::Cint, ret::API.CTypeTreeRef, args::Ptr{API.CTypeTreeRef}, known_values::Ptr{API.IntList}, numArgs::Csize_t, val::LLVM.API.LLVMValueRef)::UInt8
-    TT = TypeTree(API.DT_Integer, LLVM.context(LLVM.Value(val)))
+    # TT = TypeTree(API.DT_Integer, LLVM.context(LLVM.Value(val)))
+    #only!(TT, -1)
+    # API.EnzymeSetTypeTree(unsafe_load(args), TT)
+    # dl = string(LLVM.datalayout(LLVM.parent(LLVM.parent(LLVM.parent(LLVM.Instruction(val))))))
+    # shift!(TT,  dl, #=off=#0, #=maxSize=#8, #=addOffset=#0)
+    # API.EnzymeMergeTypeTree(TT, TypeTree(API.DT_Pointer, LLVM.context(LLVM.Value(val))))
```
Comment (vchuravy) [inline src/compiler.jl:2927]:
```suggestion
```

File: src/compiler.jl [hunk 29]
```diff
@@ -2690,11 +3245,17 @@ function create_abi_wrapper(enzymefn::LLVM.Function, F, argtypes, rettype, actua
...
             end
         elseif T <: Duplicated || T <: DuplicatedNoNeed
-            push!(T_wrapperargs, llvmT)
+            @assert width == 1
+            push!(T_wrapperargs, LLVM.LLVMType(API.EnzymeGetShadowType(width, llvmT)))
+        elseif T <: BatchDuplicated || T <: BatchDuplicatedNoNeed
+            push!(T_wrapperargs, LLVM.LLVMType(API.EnzymeGetShadowType(width, llvmT)))
+            # for _ in 1:width
+            #     push!(T_wrapperargs, llvmT)
+            # end
```
Comment (vchuravy) [inline src/compiler.jl:3258]:
```suggestion
```

File: src/compiler.jl [hunk 30]
```diff
@@ -2737,11 +3298,14 @@ function create_abi_wrapper(enzymefn::LLVM.Function, F, argtypes, rettype, actua
...
-            if rettype <: Duplicated
+            if rettype <: Duplicated || rettype <: BatchDuplicated
                 push!(T_JuliaSRet, llvmT)
             end
             if !(rettype <: Const)
-                push!(T_JuliaSRet, llvmT)
+                push!(T_JuliaSRet, LLVM.LLVMType(API.EnzymeGetShadowType(width, llvmT)))
+                # for _ in 1:width
+                #     push!(T_JuliaSRet, llvmT)
+                # end
```
Comment (vchuravy) [inline src/compiler.jl:3308]:
```suggestion
```

File: src/compiler.jl [hunk 31]
```diff
@@ -2813,6 +3377,19 @@ function create_abi_wrapper(enzymefn::LLVM.Function, F, argtypes, rettype, actua
...
+                val = params[i]
+                if isboxed
+                  val = load!(builder, val)
+                end
+                i += 1
+                #val = UndefValue(LLVM.LLVMType(API.EnzymeGetShadowType(width, llvmtype(params[i]))))
+                # for idx in 1:width
+                #    val = insert_value!(builder, val, params[i], idx-1)
+                #    i += 1
+                # end
```
Comment (vchuravy) [inline src/compiler.jl:3391]:
```suggestion
```

**Review by vchuravy — CHANGES_REQUESTED** [review]
Needs docs for new API functions e.g. `gradient` and `jacobian`.

Also can we split out the bug fixes into separate PRs? Or at least one PR with the bug fixes that is backportable to the 0.9 release?

File: test/runtests.jl (hunk 8 above)
Comment (wsmoses) [reply test/runtests.jl:852]: Sure, but a distincit PR.

File: src/compiler.jl (hunk 21 above)
Comment (wsmoses) [reply src/compiler.jl:968]: Agreed, make gradientutils.jl in subsequent PR?

File: src/compiler.jl (hunk 22 above)
Comment (wsmoses) [reply src/compiler.jl:1245]: Sure, for here -- though this part is under inspection per 1.8, which I'm going to defer to a later PR.

File: src/Enzyme.jl [hunk 32]
```diff
@@ -63,29 +64,72 @@ struct Duplicated{T} <: Annotation{T}
...
+the original result and only compute the derivative values.
+"""
 struct DuplicatedNoNeed{T} <: Annotation{T}
     val::T
     dval::T
 end
 
+
+"""
+    struct BatchDuplicated{T} <: Annotation{T}
```
Comment (vchuravy) [inline src/Enzyme.jl:82]:
```suggestion
    BatchDuplicated(x, ∂f_∂xs)
```

File: src/Enzyme.jl [hunk 33]
```diff
@@ -63,29 +64,72 @@ struct Duplicated{T} <: Annotation{T}
     dval::T
 end
 
+"""
+    struct DuplicatedNoNeed{T} <: Annotation{T}
+
+Constructor: `DuplicatedNoNeed(x, ∂f_∂x)`
```
Comment (vchuravy) [inline src/Enzyme.jl:70]:
```suggestion
    DuplicatedNoNeed(x, ∂f_∂x)

```

**Review by vchuravy — APPROVED** [review]
Will need to add a documenter preview

File: src/Enzyme.jl (hunk 20 above)
Comment (vchuravy) [reply src/Enzyme.jl:798]: alternatives:

`jacobian_reverse` or `jacobian_with_reverse`

`jacobian(::Reverse, f, x)`

Where:
```
abstract type Mode end
struct Reverse <: Mode end
struct Forward <: Mode end
```
